src/pm-conversations/pmConversationsModel.js

```diff
@@ -116,10 +116,21 @@ function insert(
   } else {
     // The conversation needs to be updated.
     map = map.set(key, msgId);
+
     const i = sorted.indexOf(key);
     invariant(i >= 0, 'pm-conversations: key in map should be in sorted');
-    sorted = sorted.delete(i); // linear time, ouch
-    return { map, sorted: insertSorted(sorted, map, key, msgId) };
+    if (i === 0) {
+      // The conversation was already the latest, so no reordering needed.
+      // (This is likely a common case in practice -- happens every time
+      // the user gets several PMs in a row in the same thread -- so good to
+      // optimize.)
+    } else {
+      // It wasn't the latest.  Just handle the general case.
+      sorted = sorted.delete(i); // linear time, ouch
+      sorted = insertSorted(sorted, map, key, msgId);
```

Comment:
Yeah, in general both of these lines can take linear time.

The common case, though, should be that the message that just came in is the newest message we've ever seen. In fact, for EVENT_NEW_MESSAGE, if it's ever *not* the newest then something odd has happened. (This could potentially happen if we go fetch the messages in some narrow, and right around the same time someone sends a message into that narrow, and the fetch sees that message and returns sooner than we get the reply from the event queue to tell us about the message. Or a similar scenario with two different messages sent right around the time we fetch: one of them in the fetch results, and a slightly older one that comes as an event shortly after. But I think that kind of race condition is basically the only way it can happen.) Meanwhile for messages we learn about through REALM_INIT or MESSAGE_FETCH_COMPLETE, if we want to optimize those they probably won't be going through this codepath anyway.

If it is the newest message we've ever seen, then the optimization described at the top of `insertSorted` applies, and it'll take constant time.


> At least in cases where we do a `.delete` followed by an `.insert`, I wonder if it would be just a bit more efficient to do a single [`.update`](https://immutable-js.github.io/immutable-js/docs/#/List/update) instead; what do you think?

The trouble is that this bit of code's job (the stanza manipulating `sorted` inside this "The conversation needs to be updated" case) is to rearrange the order of the list's elements. The conversation was at some position in the list, and we need to move it to the new position it belongs at. In general those are two different positions -- which also means we need to change which values appear at all the positions in between. So an `.update` won't be enough.

It can happen that we end up doing a `.delete` followed by `.insert` at the same position. In that case an `.update` would indeed be enough... but we'd be finding the place where this conversation's key appears, and updating it to write the conversation's key there, i.e. the same value that's already there. So better yet in that case is to do nothing.

That's what this commit toward the end of the branch is about:
16d2d7bc8 recent-pms: Optimize the case of a new message in the latest thread.
It does it for the case where we'd be doing that at position 0, because that's expected to be common. To do it in general would take more logic -- and for this to happen at any other position would mean we're seeing a message that isn't the newest message we've ever seen, which is rare, see above, so it isn't worth adding complexity for.

I'll add some more comments trying to explain all this more than the code currently does.

Reply:
> The trouble is that this bit of code's job (the stanza manipulating `sorted` inside this "The conversation needs to be updated" case) is to rearrange the order of the list's elements.

Ohh, OK, right.

> I'll add some more comments trying to explain all this more than the code currently does.

I think it's pretty well explained as-is; I just missed the mark here, oops. 🙂 Thanks for explaining. Looking back, I think I may have seen the `i` in `insert` and the `i` in `insertSorted` (totally reasonable to give them the same name), and short-circuited into thinking they'd end up being the same value, when of course they usually won't be, as you've said.